lib/linked_list.rb

```diff
@@ -12,150 +12,265 @@ def initialize(value, next_node = nil)
-    end
+    return max
+  end
 
-    # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def visit
-      raise NotImplementedError
-    end
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def length
+    return 0 if !@head
+    current = @head
+    count = 0
 
-    # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def delete(value)
-      raise NotImplementedError
+    while current
+      count += 1
+      current = current.next
     end
 
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def reverse
-      raise NotImplementedError
+    return count
+  end
+
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n), where n is the value of index
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    return if !@head
+    current = @head
+    count = 0
+  
+    until count == index
+      return nil if !current
+      current = current.next
+      count += 1
     end
 
+    return current.data
+  end
 
-    ## Advanced Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_middle_value
-      raise NotImplementedError
-    end
+  # method to print all the values in the linked list
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def visit
+    return if !@head
+    current = @head
 
-    # find the nth node from the end and return its value
-    # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_nth_from_end(n)
-      raise NotImplementedError
+    while current
+      p current.data
+      current = current.next
     end
+    
+    return nil
+  end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def has_cycle
-      raise NotImplementedError
+  # method to delete the first node found with specified value
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def delete(value)
+    return if !@head
+    current = @head
+    previous = nil
+
+    while current
+      if current.data == value
+        @head = current.next if current == @head
+        previous.next = current.next if previous # relink list if deleted value is in the middle
+        current = nil

```

Comment:
I don't think this line is actually needed as current is a local variable.

Reply:
You're right! I don't remember why I added this line.

Reply:
You're right! I don't remember why I added this line.